api/simplehash/schemav2.go

```diff
@@ -0,0 +1,198 @@
+	Behaviour          string         `json:"behaviour"`
+	TimestampDeclared  string         `json:"timestamp_declared"`
+	TimestampAccepted  string         `json:"timestamp_accepted"`
+	TimestampCommitted string         `json:"timestamp_committed"`
+	PrincipalAccepted  map[string]any `json:"principal_accepted"`
+	PrincipalDeclared  map[string]any `json:"principal_declared"`
+	ConfirmationStatus string         `json:"confirmation_status"`
+	From               string         `json:"from"`
+	TenantIdentity     string         `json:"tenant_identity"`
+}
+
+// NewEventMarshaler creates a flat marshaler to transform events to api format.
+//
+// otherwise attributes look like this: {"foo":{"str_val": "bar"}} instead of {"foo": "bar"}
+// this mimics the public list events api response, so minimises changes to the
+// public api response, to reproduce the anchor
+func NewEventMarshaler() *simpleoneof.Marshaler {
+	return v2assets.NewFlatMarshalerForEvents()
+}
+
+// TransformOne transforms a single event in grpc proto format (message bus
+// compatible) to the canonical, publicly verifiable, api format.
+func TransformOne(marshaler *simpleoneof.Marshaler, event *v2assets.EventResponse) (EventV2, error) {
+	eventJson, err := marshaler.Marshal(event)
+	if err != nil {
+		return EventV2{}, err
+	}
+	eventShashV2 := EventV2{}
+	err = json.Unmarshal(eventJson, &eventShashV2)
+	if err != nil {
+		return EventV2{}, err
+	}
+	return eventShashV2, nil
+}
+
+// PublicFromPermissionedEvent translates the permissioned event and asset identities to
+// their public counter parts.
+func PublicFromPermissionedEvent(event *v2assets.EventResponse) {
+	event.Identity = v2assets.PublicIdentityFromPermissioned(event.Identity)
+	event.AssetIdentity = v2assets.PublicIdentityFromPermissioned(event.AssetIdentity)
+}
+
+// EventSimpleHashV2 hashes a single event according to the public event format
+// available to api consumers.
+//
+//   - If the event is the permissioned (owner) counter part of a public
+//     attestation, you must call PublicFromPermissionedEvent first.
+//   - No special treatment is given to confirmation status (PENDING vs
+//     CONFIRMED). Because the rules for forestrie and PENDING events are *NOT
+//     THE SAME* as those for proof_mechanism simplehash.
+func EventSimpleHashV2(hasher hash.Hash, marshaler *simpleoneof.Marshaler, event *v2assets.EventResponse) error {
+
+	var err error
+
+	// Note that we _don't_ take any notice of confirmation status.
+
+	eventShashV2, err := TransformOne(marshaler, event)
+	if err != nil {
+		return err
+	}
+
+	// XXX: TODO I don't think the following step is necessary (we should get snake case due to the struct tags)
+	//    we get the correct fields by the definition of our structure, but we need to marshal and unmarshal our struct
+	//    into a generic []any, in order to get the correct field names, otherwise they would be camelcase
+	eventJson, err := json.Marshal(eventShashV2)
```

Comment:
Yeh I am not a big fan of 2 rounds of json marshalling/unmarshalling.

so we go from a `v2assets.EventResponse` -> `simplehash.EventV2` -> `interface{}` and the second round is just to ensure the field names are snake case? i vaguely remember that.

Reply:
Yeah, I had understood the `json:aaa_bbb` style struct tags took care of that but I can see that the attributes.* are a problem. I will take another look at this later, maybe we just want to bite the bulet and do a Marshal method. I'm going to defer that as TD I just inherited tho